### PR TITLE
feat: add consumer group

### PR DESCRIPTION
**Pull request by kingluo**
### Description

add consumer group

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/core/ctx.lua [hunk 1]
```diff
@@ -207,6 +207,7 @@ do
         balancer_ip = true,
         balancer_port = true,
         consumer_name = true,
+        consumer_group_id = true,
```
Comment (spacewander) [inline apisix/core/ctx.lua:210]: Why should we expose it as a variable? It seems we also use `ctx.consumer.group_id` in some places.

File: apisix/admin/consumer_group.lua [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    end
+
+    if need_id and conf.id and tostring(conf.id) ~= tostring(id) then
+        return nil, {error_msg = "wrong id"}
+    end
+
+    conf.id = id
+
+    core.log.info("conf: ", core.json.delay_encode(conf))
+    local ok, err = core.schema.check(core.schema.plugin_config, conf)
```
Comment (spacewander) [inline apisix/admin/consumer_group.lua:52]: Don't hijack the schema of plugin_config in other places.

File: apisix/admin/consumer_group.lua [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    local ok, err = schema_plugin(conf.plugins)
+    if not ok then
+        return nil, {error_msg = err}
+    end
+
+    return true
+end
+
+
+function _M.put(id, conf)
```
Comment (spacewander) [inline apisix/admin/consumer_group.lua:66]: We should also add a check in consumers. Let's make sure `consumer.group_id` points to an actual consumer group.

File: apisix/consumer_group.lua [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+local _M = {
+}
+
+
+function _M.init_worker()
+    local err
+    consumer_groups, err = core.config.new("/consumer_groups", {
+        automatic = true,
+        item_schema = nil,
```
Comment (spacewander) [inline apisix/consumer_group.lua:33]: Why don't we use schema to check the item?

File: t/admin/consumer-group.t [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+log_level("info");
+
+add_block_preprocessor(sub {
+    my ($block) = @_;
+
+    if (!$block->request) {
+        $block->set_value("request", "GET /t");
+    }
+
+    if (!$block->no_error_log) {
```
Comment (spacewander) [inline t/admin/consumer-group.t:32]: It is more common to use this style:
https://github.com/apache/apisix/blob/3bbc7aefe36fa0bbc0748335d64a3ead02c53f87/t/admin/filter.t#L44

File: apisix/core/ctx.lua (hunk 1 above)
Comment (kingluo) [reply apisix/core/ctx.lua:210]: We would use this variable in plugins:

For example:
```
curl http://127.0.0.1:9080/apisix/admin/consumer_groups/company_a -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "plugins": {
        "quota": {
            "rejected_code": 503,
            "minute": 2,
            "group": "$consumer_group_id"
        }
    }
}'
```

File: apisix/plugin.lua [hunk 6]
```diff
@@ -651,20 +665,33 @@ local function merge_consumer_route(route_conf, consumer_conf)
...
     core.log.info("route conf: ", core.json.delay_encode(route_conf))
     core.log.info("consumer conf: ", core.json.delay_encode(consumer_conf))
+    core.log.info("consumer group conf: ", core.json.delay_encode(consumer_group_conf))
 
     local flag = route_conf.value.id .. "#" .. route_conf.modifiedIndex
                  .. "#" .. consumer_conf.id .. "#" .. consumer_conf.modifiedIndex
+
+    if consumer_group_conf then
+        flag = flag .. "#" .. consumer_group_conf.value.id
+            .. "#" .. consumer_group_conf.modifiedIndex
```
Comment (tzssangglass) [inline apisix/plugin.lua:678]: flag should have `route_conf.value.id` and `route_conf.modifiedIndex`?

File: t/admin/consumer-group.t [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+                            "limit-count": {
+                                "count": 2,
+                                "time_window": 60,
+                                "rejected_code": 503,
+                                "key": "remote_addr"
+                            }
+                        }
+                    },
+                    "key": "/apisix/consumer_groups/company_a"
+                }]]
```
Comment (tzssangglass) [inline t/admin/consumer-group.t:71]: don't need this.

File: t/admin/consumer-group.t [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+                        "plugins": {
+                            "limit-count": {
+                                "count": 2,
+                                "time_window": 60,
+                                "rejected_code": 503,
+                                "key": "remote_addr"
+                            }
+                        }
+                    },
+                    "key": "/apisix/consumer_groups/company_a"
```
Comment (tzssangglass) [inline t/admin/consumer-group.t:248]: ditto

File: apisix/plugin.lua (hunk 6 above)
Comment (kingluo) [reply apisix/plugin.lua:678]: I already does. It concats new identification to the exist flag, which already contains route stuff.

```lua
    -- check here...
    local flag = route_conf.value.id .. "#" .. route_conf.modifiedIndex
                 .. "#" .. consumer_conf.id .. "#" .. consumer_conf.modifiedIndex

    if consumer_group_conf then
        flag = flag .. "#" .. consumer_group_conf.value.id
            .. "#" .. consumer_group_conf.modifiedIndex
    end

```

File: t/admin/consumer-group.t [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+            ngx.sleep(0.3)
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/consumer_groups/company_a',
+                 ngx.HTTP_DELETE
+            )
+            ngx.say(body)
+        }
+    }
+--- response_body
+passed
```
Comment (tzssangglass) [inline t/admin/consumer-group.t:523]: 1. we need to verify that the plugin works well in the consumer_group
2. we need to verify that consumer_group and consumer work well together

File: t/admin/consumer-group.t (hunk 7 above)
Comment (kingluo) [reply t/admin/consumer-group.t:71]: It is based on the `t/admin/plugin-configs.t`, where it associates plugins with plugin config for admin api test.
So what's the reason to remove them?

File: t/admin/consumer-group.t (hunk 9 above)
Comment (kingluo) [reply t/admin/consumer-group.t:523]: Just like what other resources does, e.g. plugin_config, consumer, this test file is only for admin api test.
I would put plugin test and consumer stuff in `t/node/consumer-group.t`, and moreover, `t/config-center-yaml/consumer-group.t`, please keep in patient.

File: t/admin/consumer-group.t (hunk 8 above)
Comment (tzssangglass) [reply t/admin/consumer-group.t:248]: We rarely validate return values in this way these days. Perhaps the `plugin-config.t` you referenced would be acceptable if left as is.

Comment (kingluo) [reply t/admin/consumer-group.t:248]: Then which way or which test file do you think I should take reference?

Comment (tzssangglass) [reply t/admin/consumer-group.t:248]: I get it, let's keep it.

File: apisix/core/ctx.lua (hunk 1 above)
Comment (spacewander) [inline apisix/core/ctx.lua:210]: For alphabetical order, the `consumer_group_id` is ahead of consumer_name

File: apisix/admin/consumers.lua [hunk 10]
```diff
@@ -62,6 +63,12 @@ local function check_conf(username, conf)
         end
     end
 
+    if conf.group_id then
+        if consumer_group.get(conf.group_id) == nil then
```
Comment (spacewander) [inline apisix/admin/consumers.lua:67]: For code style, I will recommend using the same solution in https://github.com/apache/apisix/blob/3c18031c6e70daa15db72fe06ae8644b04b4481e/apisix/admin/routes.lua#L79
1. fetch the data from etcd if possible
2. use a similar error message

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

File: docs/en/latest/terminology/consumer-group.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            "key": "auth-one"
+        }
+    },
+    "group_id": "company_a"
+}'
+```
+
+When APISIX can't find the Consumer Group with the `group_id`, the Admin API is terminated with a status code of 400.
+
+If a Consumer already has the `plugins` field configured, the plugins in the Consumer Group will effectively be merged to it. The same plugin in the Consumer Group will not override the ones configured directly in the Consumer.
```
Comment (spacewander) [inline docs/en/latest/terminology/consumer-group.md:68]:
```suggestion
If a Consumer already has the `plugins` field configured, the plugins in the Consumer Group will effectively be merged into it. The same plugin in the Consumer Group will not override the one configured directly in the Consumer.
```

File: docs/en/latest/terminology/consumer-group.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+#
+-->
+
+Consumer Groups are used to extract commonly used [Plugin](./plugin.md) configurations and can be bound directly to a [Consumer](./consumer.md).
+
+With consumer groups, you can define any number of plugins, e.g. rate limiting and apply them to a set of consumers,
+instead of managing each consumer individually.
+
+While configuring the same plugin for the same route, only one copy of the configuration is valid.
+The order of precedence is `Consumer` > `Consumer Group` > `Route` > `plugin_config` > `Service`.
```
Comment (spacewander) [inline docs/en/latest/terminology/consumer-group.md:35]: Better to put the order to one place: https://github.com/apache/apisix/blob/61afafdd46fe40c03955b6cb058df3e6fad5db45/docs/en/latest/terminology/plugin.md?plain=1#L30

File: docs/en/latest/terminology/consumer-group.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+}'
+```
+
+When APISIX can't find the Consumer Group with the `group_id`, the Admin API is terminated with a status code of 400.
+
+If a Consumer already has the `plugins` field configured, the plugins in the Consumer Group will effectively be merged to it. The same plugin in the Consumer Group will not override the ones configured directly in the Consumer.
+
+For example, if we configure a Consumer Group as shown below
+
+```
```
Comment (spacewander) [inline docs/en/latest/terminology/consumer-group.md:72]: Let's add language mark like other examples

File: docs/en/latest/terminology/consumer-group.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+Consumer Groups are used to extract commonly used [Plugin](./plugin.md) configurations and can be bound directly to a [Consumer](./consumer.md).
```
Comment (spacewander) [inline docs/en/latest/terminology/consumer-group.md:29]: We should update the admin-api.md?

File: apisix/core/ctx.lua [hunk 15]
```diff
@@ -206,6 +206,7 @@ do
     local apisix_var_names = {
         balancer_ip = true,
         balancer_port = true,
+        consumer_group_id = true,
```
Comment (spacewander) [inline apisix/core/ctx.lua:209]: Need to update https://github.com/apache/apisix/blob/master/docs/en/latest/apisix-variable.md for the new variable.

File: docs/en/latest/terminology/consumer-group.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    },
+    "group_id": "company_a"
+}'
+```
+
+When APISIX can't find the Consumer Group with the `group_id`, the Admin API is terminated with a status code of 400.
+
+If a Consumer already has the `plugins` field configured, the plugins in the Consumer Group will effectively be merged to it. The same plugin in the Consumer Group will not override the ones configured directly in the Consumer.
+
+For example, if we configure a Consumer Group as shown below
```
Comment (SylviaBABY) [inline docs/en/latest/terminology/consumer-group.md:70]: Add `.` or `:` in the end of sentence

File: docs/en/latest/terminology/consumer-group.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    "id": "bar",
+    "plugins": {
+        "response-rewrite": {
+            "body": "hello"
+        }
+    }
+}
+```
+
+to a Consumer as shown below,
```
Comment (SylviaBABY) [inline docs/en/latest/terminology/consumer-group.md:83]:
```suggestion
To a Consumer as shown below.
```

File: docs/en/latest/terminology/consumer-group.md [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+```
```
Comment (SylviaBABY) [inline docs/en/latest/terminology/consumer-group.md:85]: need to add code language too

File: docs/en/latest/terminology/consumer-group.md [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    "plugins": {
+        "key-auth": {
+            "key": "auth-one"
+        }
+    },
+    "group_id": "company_a"
+}'
+```
+
+When APISIX can't find the Consumer Group with the `group_id`, the Admin API is terminated with a status code of 400.
```
Comment (SylviaBABY) [inline docs/en/latest/terminology/consumer-group.md:66]:
```suggestion
When APISIX can't find the Consumer Group with the `group_id`, the Admin API is terminated with a status code of `400`.
```

File: docs/en/latest/apisix-variable.md [hunk 20]
```diff
@@ -38,6 +38,7 @@ additional variables.
 | balancer_ip         | core       | The IP of picked upstream server.                                                   | 192.168.1.2    |
 | balancer_port       | core       | The port of picked upstream server.                                                 | 80             |
 | consumer_name       | core       | Username of Consumer.                                                               |                |
+| consumer_group_id   | core       | Group name of Consumer.                                                             |                |
```
Comment (spacewander) [inline docs/en/latest/apisix-variable.md:41]:
```suggestion
| consumer_group_id   | core       | Group ID of Consumer.                                                             |                |
```

File: docs/en/latest/admin-api.md [hunk 21]
```diff
@@ -947,6 +948,35 @@ Sets Plugins which run globally. i.e these Plugins will be run before any Route/
 | create_time | False    | Epoch timestamp (in seconds) of the created time. If missing, this field will be populated automatically.             | 1602883670 |
 | update_time | False    | Epoch timestamp (in seconds) of the updated time. If missing, this field will be populated automatically.             | 1602883670 |
 
+## Consumer group
```
Comment (spacewander) [inline docs/en/latest/admin-api.md:951]: Let's update the `Resources that support paging queries:` list in this page

File: docs/zh/latest/apisix-variable.md [hunk 22]
```diff
@@ -37,6 +37,7 @@ APISIX 除了支持 [NGINX 变量](http://nginx.org/en/docs/varindex.html)外，
 | balancer_ip         | core       | 上游服务器的 IP 地址。                                                            | 192.168.1.2      |
 | balancer_port       | core       | 上游服务器的端口。                                                                | 80               |
 | consumer_name       | core       | 消费者的名称。                                                                    |                  |
+| consumer_group_id   | core       | 消费者所在的组的名称。                                                            |                  |
```
Comment (spacewander) [inline docs/zh/latest/apisix-variable.md:40]: Let's sync the change to the Chinese version too

**Review by spacewander — APPROVED** [review]

**Review by SylviaBABY — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**